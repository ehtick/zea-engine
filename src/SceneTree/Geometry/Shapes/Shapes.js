export * from './Rect.js';
export * from './Circle.js';
export * from './Grid.js';
export * from './Plane.js';
export * from './Cuboid.js';
export * from './Cylinder.js';
export * from './Cone.js';
export * from './Torus.js';
export * from './Sphere.js';

