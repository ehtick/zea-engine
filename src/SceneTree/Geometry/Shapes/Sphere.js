import { Vec2 } from '../../../Math/Vec2';
import { Vec3 } from '../../../Math/Vec3';
import { Mesh } from '../Mesh.js';


class Sphere extends Mesh {
    constructor(radius = 1.0, sides = 12, loops = 12) {
        super();

        if(isNaN(radius) || isNaN(sides) || isNaN(loops))
            throw("Invalid geom args");

        this.__radius = radius;
        this.__sides = (sides >= 3) ? sides : 3;
        this.__loops = (loops >= 3) ? loops : 3;

        this.addVertexAttribute('texCoords', Vec2);
        this.addVertexAttribute('normals', Vec3);
        this.__rebuild();
    }

    get radius() {
        return this.__radius
    }

    set radius(val) {
        this.__radius = val;
        this.__resize();
    }

    get sides() {
        return this.__sides
    }

    set sides(val) {
        this.__sides = (val >= 3) ? val : 3;
        this.__rebuild();
    }

    get loops() {
        return this.__loops
    }

    set loops(val) {
        this.__loops = (val >= 3) ? val : 3;
        this.__rebuild();
    }

    __rebuild() {
        let numVertices = 2 + this.__sides * this.__loops;
        let numTris = this.__sides * 2;
        let numQuads = this.__sides * this.__loops;
        this.setNumVertices(numVertices);
        this.setFaceCounts([numTris, numQuads]);

        //////////////////////////////
        // Set Vertex Positions

        let normals = this.getVertexAttribute('normals');
        let vertex = 0;
        let normal = new Vec3(0.0, 0.0, 1.0);
        this.getVertex(vertex).set(0.0, 0.0, this.__radius);
        normals.getValueRef(vertex).set(0.0, 0.0, 1.0);
        vertex++;

        for (let i = 0; i < this.__loops; i++) {
            let theta = ((i + 1) / (this.__loops + 1)) * Math.PI;
            for (let j = 0; j < this.__sides; j++) {
                let phi = (j / this.__sides) * 2.0 * Math.PI;
                normal.set(Math.sin(theta) * Math.cos(phi), Math.sin(theta) * Math.sin(phi), Math.cos(theta));

                // Set positions and normals at the same time.
                this.getVertex(vertex).setFromOther(normal.scale(this.__radius));
                normals.getValueRef(vertex).setFromOther(normal);
                vertex++;
            }
        }
        this.getVertex(vertex).set(0.0, 0.0, -this.__radius);
        normals.getValueRef(vertex).set(0.0, 0.0, -1.0);
        vertex++;

        //////////////////////////////
        // build the topology
        let texCoords = this.getVertexAttribute('texCoords');

        // build the fan at the first pole.
        let faceIndex = 0;
        for (let j = 0; j < this.__sides; j++) {
            let v0 = 0;
            let v1 = ((j + 1) % this.__sides) + 1;
            let v2 = j + 1;
            this.setFaceVertexIndices(faceIndex, v0, v1, v2);

            let uv0 = new Vec2(0.5, 0.0);
            let uv1 = new Vec2(1.0 - ((j + 1) / this.__sides), 0.0);
            let uv2 = new Vec2(1.0 - (j / this.__sides), 1.0 / (this.__loops + 1));
            texCoords.setFaceVertexValue(faceIndex, 0, uv0);
            texCoords.setFaceVertexValue(faceIndex, 1, uv1);
            texCoords.setFaceVertexValue(faceIndex, 2, uv2);

            faceIndex++;
        }
        // build the fan at the second pole.
        for (let j = 0; j < this.__sides; j++) {
            let v0 = numVertices - 1;
            let v1 = (this.__sides * (this.__loops - 1)) + j + 1;
            let v2 = (this.__sides * (this.__loops - 1)) + ((j + 1) % this.__sides) + 1;
            this.setFaceVertexIndices(faceIndex, v0, v1, v2);

            let uv0 = new Vec2(1.0 - (j / this.__sides), this.__loops / (this.__loops + 1));
            let uv1 = new Vec2(1.0 - ((j + 1) / this.__sides), this.__loops / (this.__loops + 1));
            let uv2 = new Vec2(0.5, 1.0);
            texCoords.setFaceVertexValue(faceIndex, 0, uv0);
            texCoords.setFaceVertexValue(faceIndex, 1, uv1);
            texCoords.setFaceVertexValue(faceIndex, 2, uv2);

            faceIndex++;
        }

        for (let i = 0; i < this.__loops - 1; i++) {
            for (let j = 0; j < this.__sides; j++) {
                let v0 = (this.__sides * i) + j + 1;
                let v1 = (this.__sides * i) + ((j + 1) % this.__sides) + 1;
                let v2 = (this.__sides * (i + 1)) + ((j + 1) % this.__sides) + 1;
                let v3 = (this.__sides * (i + 1)) + j + 1;
                this.setFaceVertexIndices(faceIndex, v0, v1, v2, v3);

                texCoords.setFaceVertexValue(faceIndex, 0, new Vec2(i / this.__loops, j / this.__loops));
                texCoords.setFaceVertexValue(faceIndex, 1, new Vec2(i / this.__loops, (j + 1) / this.__loops));
                texCoords.setFaceVertexValue(faceIndex, 2, new Vec2((i + 1) / this.__loops, (j + 1) / this.__loops));
                texCoords.setFaceVertexValue(faceIndex, 3, new Vec2((i + 1) / this.__loops, j / this.__loops));
                faceIndex++;
            }
        }

        this.setBoundingBoxDirty();
    }

    __resize() {
        //////////////////////////////
        // Set Vertex Positions
        let vertex = 0;
        let normal = new Vec3(0.0, 0.0, 1.0);
        this.getVertex(vertex).set(0.0, 0.0, this.__radius);
        vertex++;

        for (let i = 0; i < this.__loops; i++) {
            let theta = ((i + 1) / (this.__loops + 1)) * Math.PI;
            for (let j = 0; j < this.__sides; j++) {
                let phi = (j / this.__sides) * 2.0 * Math.PI;
                normal.set(Math.sin(theta) * Math.cos(phi), Math.sin(theta) * Math.sin(phi), Math.cos(theta));

                // Set positions and normals at the same time.
                this.getVertex(vertex).setFromOther(normal.scale(this.__radius));
                vertex++;
            }
        }
        this.getVertex(vertex).set(0.0, 0.0, -this.__radius);
        vertex++;

        this.setBoundingBoxDirty();
    }

    toJSON() {
        let json = super.toJSON();
        json['x'] = this.__x;
        json['y'] = this.__y;
        json['z'] = this.__z;
        return json
    }
};

export {
    Sphere
};
//export default Sphere;