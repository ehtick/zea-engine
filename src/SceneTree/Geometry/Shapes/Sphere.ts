import { Vec2, Vec3 } from '../../../Math/index'
import { NumberParameter } from '../../Parameters/NumberParameter'
import { Registry } from '../../../Registry'
import { ProceduralMesh } from './ProceduralMesh'
import { Vec3Attribute } from '../Vec3Attribute'
import { Vec2Attribute } from '../Vec2Attribute'

/**
 * A class for generating a sphere geometry.
 *
 * ```
 * const sphere = new Sphere(1.4, 13)
 * ```
 *
 * **Parameters**
 * * **Radius(`NumberParameter`):** Radius of the sphere.
 * * **Sides(`NumberParameter`):** Specifies the number of subdivisions around the `Z` axis.
 * * **Loops(`NumberParameter`):** Specifies the number of subdivisions(stacks) along the `Z` axis.
 *
 * @extends {ProceduralMesh}
 */
class Sphere extends ProceduralMesh {
  /**
   * @member {NumberParameter} loopsParam - Specifies the number of subdivisions(stacks) along the `Z` axis.
   */
  loopsParam: NumberParameter

  /**
   * @member {NumberParameter} radiusParam - Radius of the sphere.
   */
  radiusParam: NumberParameter

  /**
   * @member {NumberParameter} sidesParam - Specifies the number of subdivisions around the `Z` axis.
   */
  sidesParam: NumberParameter

  /**
   * Creates an instance of Sphere.
   * @param {number} [radius=1.0] - The radius of the sphere.
   * @param {number} [sides=12] - The number of sides.
   * @param {number} [loops=12] - The number of loops.
   * @param {boolean} [addNormals=true] - Compute vertex normals for the geometry
   * @param {boolean} [addTextureCoords=true] - Compute texture coordinates for the geometry
   */
  constructor(radius = 1.0, sides = 12, loops = 12) {
    super()

    if (isNaN(radius) || isNaN(sides) || isNaN(loops)) throw new Error('Invalid geom args')

    this.radiusParam = this.addParameter(new NumberParameter('Radius', radius)) as NumberParameter
    this.sidesParam = this.addParameter(
      new NumberParameter('Sides', sides >= 3 ? sides : 3, [3, 200], 1)
    ) as NumberParameter
    this.loopsParam = this.addParameter(
      new NumberParameter('Loops', loops >= 3 ? loops : 3, [3, 200], 1)
    ) as NumberParameter

    this.addVertexAttribute('texCoords', new Vec2Attribute())
    this.addVertexAttribute('normals', new Vec3Attribute())

    this.topologyParams.push('Sides')
    this.topologyParams.push('Loops')
  }

  /**
   * The rebuild method.
   * @private
   */
  rebuild(): void {
    const radius = this.radiusParam.value || 1.0
    const nbSides = this.sidesParam.value || 12
    const nbLoops = this.loopsParam.value || 12

    const numVertices = 2 + nbSides * nbLoops
    const numTris = nbSides * 2
    const numQuads = nbSides * nbLoops
    this.setNumVertices(numVertices)
    this.setFaceCounts([numTris, numQuads])

    // ////////////////////////////
    // Set Vertex Positions

    const positions = <Vec3Attribute>this.getVertexAttribute('positions')
    const normals = <Vec3Attribute>this.getVertexAttribute('normals')
    const normal = new Vec3(0.0, 0.0, 1.0)
    let vertex = 0
    if (!positions || !normals) return

    positions.getValueRef(vertex).set(0.0, 0.0, radius)
    normals.getValueRef(vertex).set(0.0, 0.0, 1.0)
    vertex++

    for (let i = 0; i < nbLoops; i++) {
      const theta = ((i + 1) / (nbLoops + 1)) * Math.PI
      for (let j = 0; j < nbSides; j++) {
        const phi = -((j / nbSides) * 2.0 * Math.PI)
        normal.set(Math.sin(theta) * Math.cos(phi), Math.sin(theta) * Math.sin(phi), Math.cos(theta))

        // Set positions and normals at the same time.
        positions.getValueRef(vertex).setFromOther(normal.scale(radius))
        normals.getValueRef(vertex).setFromOther(normal)
        vertex++
      }
    }

    positions.getValueRef(vertex).set(0.0, 0.0, -radius)
    normals.getValueRef(vertex).set(0.0, 0.0, -1.0)
    vertex++

    // ////////////////////////////
    // Build the topology
    const texCoords = <Vec2Attribute>this.getVertexAttribute('texCoords')
    if (texCoords) {
      // build the fan at the first pole.
      let faceIndex = 0
      for (let j = 0; j < nbSides; j++) {
        const v0 = 0
        const v1 = ((j + 1) % nbSides) + 1
        const v2 = j + 1
        this.setFaceVertexIndices(faceIndex, [v0, v1, v2])

        if (texCoords) {
          const uv0 = new Vec2(0.5, 0.0)
          const uv1 = new Vec2(1.0 - (j + 1) / nbSides, 0.0)
          const uv2 = new Vec2(1.0 - j / nbSides, 1.0 / (nbLoops + 1))
          texCoords.setFaceVertexValue(faceIndex, 0, uv0)
          texCoords.setFaceVertexValue(faceIndex, 1, uv1)
          texCoords.setFaceVertexValue(faceIndex, 2, uv2)
        }

        faceIndex++
      }

      // Build the fan at the second pole.
      for (let j = 0; j < nbSides; j++) {
        const v0 = numVertices - 1
        const v1 = nbSides * (nbLoops - 1) + j + 1
        const v2 = nbSides * (nbLoops - 1) + ((j + 1) % nbSides) + 1
        this.setFaceVertexIndices(faceIndex, [v0, v1, v2])

        const uv0 = new Vec2(1.0 - j / nbSides, nbLoops / (nbLoops + 1))
        const uv1 = new Vec2(1.0 - (j + 1) / nbSides, nbLoops / (nbLoops + 1))
        const uv2 = new Vec2(0.5, 1.0)
        texCoords.setFaceVertexValue(faceIndex, 0, uv0)
        texCoords.setFaceVertexValue(faceIndex, 1, uv1)
        texCoords.setFaceVertexValue(faceIndex, 2, uv2)

        faceIndex++
      }

      for (let i = 0; i < nbLoops - 1; i++) {
        for (let j = 0; j < nbSides; j++) {
          const v0 = nbSides * i + j + 1
          const v1 = nbSides * i + ((j + 1) % nbSides) + 1
          const v2 = nbSides * (i + 1) + ((j + 1) % nbSides) + 1
          const v3 = nbSides * (i + 1) + j + 1
          this.setFaceVertexIndices(faceIndex, [v0, v1, v2, v3])

          texCoords.setFaceVertexValue(faceIndex, 0, new Vec2(i / nbLoops, j / nbLoops))
          texCoords.setFaceVertexValue(faceIndex, 1, new Vec2(i / nbLoops, (j + 1) / nbLoops))
          texCoords.setFaceVertexValue(faceIndex, 2, new Vec2((i + 1) / nbLoops, (j + 1) / nbLoops))
          texCoords.setFaceVertexValue(faceIndex, 3, new Vec2((i + 1) / nbLoops, j / nbLoops))
          faceIndex++
        }
      }
    }
  }

  /**
   * The resize method.
   * @private
   */
  resize() {
    const radius = this.radiusParam.value
    const nbSides = this.sidesParam.value
    const nbLoops = this.loopsParam.value
    if (!radius || !nbSides || !nbLoops) {
      console.warn('resize() failed')
      return
    }
    // ////////////////////////////
    // Set Vertex Positions
    const positions = <Vec3Attribute>this.getVertexAttribute('positions')
    const normals = <Vec3Attribute>this.getVertexAttribute('normals')
    let vertex = 0
    const normal = new Vec3(0.0, 0.0, 1.0)
    positions.getValueRef(vertex).set(0.0, 0.0, radius)
    if (normals) normals.getValueRef(vertex).set(0.0, 0.0, 1.0)
    vertex++

    for (let i = 0; i < nbLoops; i++) {
      const theta = ((i + 1) / (nbLoops + 1)) * Math.PI
      for (let j = 0; j < nbSides; j++) {
        const phi = -((j / nbSides) * 2.0 * Math.PI)
        normal.set(Math.sin(theta) * Math.cos(phi), Math.sin(theta) * Math.sin(phi), Math.cos(theta))

        // Set positions and normals at the same time.
        positions.getValueRef(vertex).setFromOther(normal.scale(radius))
        if (normals) normals.getValueRef(vertex).setFromOther(normal)
        vertex++
      }
    }
    positions.getValueRef(vertex).set(0.0, 0.0, -radius)
    if (normals) normals.getValueRef(vertex).set(0.0, 0.0, -1.0)
    vertex++
  }
}

Registry.register('Sphere', Sphere)

export { Sphere }
