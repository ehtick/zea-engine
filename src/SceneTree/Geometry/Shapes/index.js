export * from './PointGrid.js';

export * from './Rect.js';
export * from './Circle.js';
export * from './Cross.js';
export * from './LinesCuboid.js';
export * from './Grid.js';

export * from './Cone.js';
export * from './Cuboid.js';
export * from './Cylinder.js';
export * from './Disc.js';
export * from './Plane.js';
export * from './Sphere.js';
export * from './Torus.js';
