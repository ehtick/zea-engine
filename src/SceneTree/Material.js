import {
    Vec2,
    Vec3,
    Color
} from '../Math';
import {
    Signal
} from '../Utilities';
import {
    BaseItem
} from './BaseItem.js';
import {
    Image2D
} from './Image2D.js';
import {
    RefCounted
} from './RefCounted.js';
import {
    Parameter,
    NumberParameter,
    Vec2Parameter,
    Vec3Parameter,
    ColorParameter
} from './Parameters';

let makeParameterTexturable = (parameter) => {
    let image = undefined;
    parameter.textureConnected = new Signal();
    parameter.textureDisconnected = new Signal();

    let basegetValue = parameter.getValue;
    parameter.getValue = (getTexIfAvailable = true) => {
        if (getTexIfAvailable && image != undefined)
            return image;
        else
            return basegetValue();
    }
    parameter.getImage = () => {
        return image;
    }

    let imageUpdated = () => {
        parameter.valueChanged.emit(image);
    }
    
    parameter.setImage = (value) => {
        let disconnectImage = () => {
            image.removeRef(parameter);
            image.loaded.disconnect(imageUpdated);
            image.updated.disconnect(imageUpdated);
            parameter.textureDisconnected.emit();
        }
        if (value) {
            if (image != undefined && image !== value) {
                disconnectImage();
            }
            image = value;
            image.addRef(parameter);
            image.loaded.connect(imageUpdated);
            image.updated.connect(imageUpdated);
            parameter.textureConnected.emit();
            parameter.valueChanged.emit(image);
        } else {
            if (image != undefined) {
                disconnectImage();
                image = undefined;
                parameter.textureDisconnected.emit();
            }
        }
    }

    let basesetValue = parameter.setValue;
    parameter.setValue = (value) => {
        if (value instanceof Image2D) {
            parameter.setImage(value);
        } else {
            if (image != undefined) {
                parameter.setImage(undefined);
            }
            basesetValue(value);
        }
    }

    // Invoke the setter so if the value is a texture, the parmater is updated.
    parameter.setValue(basegetValue());
};


class Material extends BaseItem {
    constructor(name, shaderName) {
        super(name);
        this.__shaderName = shaderName;

        this.updated = new Signal();
        this.textureConnected = new Signal();
        this.textureDisconnected = new Signal();
        this.shaderNameChanged = new Signal();
    }

    getShaderName() {
        return this.__shaderName;
    }

    setShaderName(shaderName) {
        this.__shaderName = shaderName;
        this.shaderNameChanged.emit(this.__shaderName);
    }

    destroy() {
        this.removeAllTextures();
        super.destroy();
    }

    removeAllTextures() {
        for (let paramName in this.__params) {
            if (this.__params[paramName].getImage()) {
                this.__params[paramName].getImage().removeRef(this);
                this.__params[paramName].setImage(undefined);
            }
        }
    }

    copyFrom(srcMaterial) {
        for (let paramName in this.__params) {
            let prop = this.__params[paramName];
            let srcParam = srcMaterial.getParameter(paramName)
            if (srcParam != undefined)
                this.__params[paramName] = srcParam;
        }
    }

    ///////////////////////////////
    // Parameters

    getParamTextures() {
        let textures = {};
        for (let param of this.__params) {
            if (param.getImage())
                textures[param.getName()] = param.getImage();
        }
        return textures;
    }

    __makeParameterTexturable(param) {
        makeParameterTexturable(param);
        param.textureConnected.connect(this.textureConnected.emit);
        param.textureDisconnected.connect(this.textureDisconnected.emit);
    }

    addParameter(paramName, defaultValue) {
        let image;
        if (defaultValue instanceof Image2D) {
            image = defaultValue;
            defaultValue = new Color();
        }
        let param = super.addParameter(paramName, defaultValue);
        param.valueChanged.connect(this.updated.emit);
        if(!param.setImage)
            this.__makeParameterTexturable(param);
        if (image) {
            param.setImage(image)
        }
        return param;
    }

    addParameterInstance(param) {
        super.addParameterInstance(param);
        this.__makeParameterTexturable(param);
        return param;
    }

    isTransparent() {
        let opacity = this.getParameter('opacity');
        if (opacity && (opacity.getValue() < 0.99 || opacity.getImage()))
            return true;
        let baseColor = this.getParameter('baseColor');
        if (baseColor && baseColor.getImage() && baseColor.getImage().channels == 'RGBA')
            return true;
        return false;
    }

    //////////////////////////////////////////
    // Persistence

    toJSON() {
        return super.toJSON();
    }

    fromJSON(j) {
        super.fromJSON(j);
        let props = this.__params;
        for (let key in j) {
            let value;
            if (j[key] instanceof Object) {
                value = new Color();
                value.fromJSON(j[key]);
            } else {
                value = j[key];
            }
            this.addParameter(paramName, value);
        }
    }

    readBinary(reader, flags, textureLibrary) {
        super.readBinary(reader, flags);

        const numParams = reader.loadUInt32();
        for (let i = 0; i < numParams; i++) {
            const paramName = reader.loadStr();
            const paramType = reader.loadStr();
            let value;
            if (paramType == "MaterialColorParam") {
                value = reader.loadRGBAFloat32Color();
                // If the value is in linear space, then we should convert it to gamma space.
                // Note: !! this should always be done in preprocessing...
                value.applyGamma(2.2);
            } else {
                value = reader.loadFloat32();
            }
            // console.log(paramName +":" + value);
            const param = this.addParameter(paramName, value);
            const textureName = reader.loadStr();
            if (textureLibrary[textureName]) {
                // console.log(paramName +":" + textureName + ":" + textureLibrary[textureName].resourcePath);
                param.setValue(textureLibrary[textureName]);
            }
            else if(textureName!= ''){
                console.warn("Missing Texture:" + textureName)
            }
        }
    }

};
export {
    makeParameterTexturable,
    Material
};
// Material;