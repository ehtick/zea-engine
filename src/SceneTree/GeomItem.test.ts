import { TreeItem } from './TreeItem'
import { GeomItem } from './GeomItem'
import { Vec3, Xfo, Color, Mat4, Quat } from '../Math'
import { Material } from './Material'
import { Sphere } from './Geometry/Shapes/Sphere'
import '../Renderer/Shaders/SimpleSurfaceShader'
import Fixtures from './_fixtures_/GeomItem.fixture'

describe('GeomItem', () => {
  it('tests default parameters', () => {
    const geoItem = new GeomItem()

    expect(geoItem.geomParam.value).toBeUndefined()
    expect(geoItem.materialParam.value).toBeUndefined()
    expect(geoItem.geomOffsetXfoParam.value).toEqual(new Xfo())
    expect(geoItem.geomMatParam.value).toEqual(new Mat4())
  })

  it('updates parameters', () => {
    const geoItem = new GeomItem()
    geoItem.geomParam.value = new Sphere(1.4, 13)
    expect(geoItem.geomParam.value.toJSON()).toEqual(new Sphere(1.4, 13).toJSON())

    const standardMaterial = new Material('myMaterial', 'SimpleSurfaceShader')
    standardMaterial.getParameter('BaseColor').value = new Color(89 / 255, 182 / 255, 92 / 255)
    geoItem.materialParam.value = standardMaterial
    expect(geoItem.materialParam.value).toEqual(standardMaterial)

    const xfo = new Xfo(new Vec3(1, 2, 3), new Quat(0, 0, 1, 0), new Vec3(8, 9, 10))
    geoItem.geomOffsetXfoParam.value = xfo
    expect(geoItem.geomOffsetXfoParam.value.toJSON()).toEqual(xfo.toJSON())

    // This is computed, so the returned value is different
    const mat4 = new Mat4(-8, 0, 0, 0, 0, -9, 0, 0, 0, 0, 10, 0, 1, 2, 3, 1)
    geoItem.geomMatParam.value = mat4
    expect(geoItem.geomMatParam.value).toEqual(mat4)
  })

  test('test GeomOffsetXfo and GeomMat.', () => {
    const parent = new TreeItem('Parent')
    const child = new GeomItem('Child')
    parent.addChild(child)

    parent.localXfoParam.value = new Xfo(new Vec3(5, 2, 0))
    child.localXfoParam.value = new Xfo(new Vec3(2, 4, 0))
    child.geomOffsetXfoParam.value = new Xfo(new Vec3(2, 4, 0))

    let correctXfo = new Xfo(new Vec3(7, 6, 0), new Quat(0, 0, 0, 1))

    // {
    //   ori: { w: 1, x: 0, y: 0, z: 0 },
    //   tr: { x: 7, y: 6, z: 0 }
    // }
    expect(child.globalXfoParam.value.approxEqual(correctXfo, 0.001)).toBe(true)

    expect(child.geomMatParam.value.asArray()).toEqual(Float32Array.of(1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 9, 10, 0, 1))
  })

  test('Saving to JSON (serialization).', () => {
    const standardMaterial = new Material('myMaterial', 'SimpleSurfaceShader')
    standardMaterial.getParameter('BaseColor').value = new Color(89 / 255, 182 / 255, 92 / 255)

    const geomItem = new GeomItem('Item', new Sphere(1.4, 13), standardMaterial)
    geomItem.localXfoParam.value = new Xfo(new Vec3(2, 4, 0))
    geomItem.geomOffsetXfoParam.value = new Xfo(new Vec3(2, 4, 0))
    // To trigger a new calculation of the BBox we need to request its value
    geomItem.boundingBoxParam.value
    expect(geomItem.toJSON()).toMatchSnapshot()
  })

  it('loads from JSON (serialization).', () => {
    const geomItem = new GeomItem('Item')
    geomItem.fromJSON(Fixtures.fromJSON, {})

    const standardMaterial = new Material('myMaterial', 'SimpleSurfaceShader')
    standardMaterial.getParameter('BaseColor').value = new Color(89 / 255, 182 / 255, 92 / 255)

    const defaultItem = new GeomItem('Item', new Sphere(1.4, 13), standardMaterial)
    defaultItem.localXfoParam.value = new Xfo(new Vec3(2, 4, 0))
    defaultItem.geomOffsetXfoParam.value = new Xfo(new Vec3(2, 4, 0))
    // To trigger a new calculation of the BBox we need to request its value
    defaultItem.boundingBoxParam.value

    expect(geomItem.toJSON()).toEqual(defaultItem.toJSON())
  })

  it('clones GeomItem', () => {
    const standardMaterial = new Material('myMaterial', 'SimpleSurfaceShader')
    standardMaterial.getParameter('BaseColor').value = new Color(89 / 255, 182 / 255, 92 / 255)

    const item = new GeomItem('Item', new Sphere(1.4, 13), standardMaterial)
    item.localXfoParam.value = new Xfo(new Vec3(2, 4, 0))
    item.geomOffsetXfoParam.value = new Xfo(new Vec3(2, 4, 0))

    const clonedItem = item.clone()
    expect(item.toJSON()).toEqual(clonedItem.toJSON())
  })

  it('copies from another GeomItem', () => {
    const standardMaterial = new Material('myMaterial', 'SimpleSurfaceShader')
    standardMaterial.getParameter('BaseColor').value = new Color(89 / 255, 182 / 255, 92 / 255)

    const item = new GeomItem('Item', new Sphere(1.4, 13), standardMaterial)
    item.localXfoParam.value = new Xfo(new Vec3(2, 4, 0))
    item.geomOffsetXfoParam.value = new Xfo(new Vec3(2, 4, 0))

    const copiedItem = new GeomItem()
    copiedItem.copyFrom(item)
    expect(item.toJSON()).toEqual(copiedItem.toJSON())
  })
})
