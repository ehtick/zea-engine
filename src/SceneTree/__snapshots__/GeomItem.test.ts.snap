// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GeomItem Saving to JSON (serialization). 1`] = `
Object {
  "name": "Item",
  "params": Object {
    "BoundingBox": Object {
      "value": Object {
        "p0": Object {
          "x": 2.65059232711792,
          "y": 6.620340824127197,
          "z": -1.399999976158142,
        },
        "p1": Object {
          "x": 5.389792442321777,
          "y": 9.379659652709961,
          "z": 1.399999976158142,
        },
      },
    },
    "GeomOffsetXfo": Object {
      "name": "GeomOffsetXfo",
      "value": Object {
        "ori": Object {
          "w": 1,
          "x": 0,
          "y": 0,
          "z": 0,
        },
        "sc": Object {
          "x": 1,
          "y": 1,
          "z": 1,
        },
        "tr": Object {
          "x": 2,
          "y": 4,
          "z": 0,
        },
      },
    },
    "Geometry": Object {
      "name": "Geometry",
      "value": Object {
        "params": Object {
          "Loops": Object {
            "range": Array [
              3,
              200,
            ],
            "step": 1,
            "value": 12,
          },
          "Radius": Object {
            "value": 1.4,
          },
          "Sides": Object {
            "range": Array [
              3,
              200,
            ],
            "step": 1,
            "value": 13,
          },
        },
        "type": "Sphere",
        "vertexAttributes": Object {},
      },
    },
    "LocalXfo": Object {
      "name": "LocalXfo",
      "value": Object {
        "ori": Object {
          "w": 1,
          "x": 0,
          "y": 0,
          "z": 0,
        },
        "sc": Object {
          "x": 1,
          "y": 1,
          "z": 1,
        },
        "tr": Object {
          "x": 2,
          "y": 4,
          "z": 0,
        },
      },
    },
    "Material": Object {
      "value": Object {
        "name": "myMaterial",
        "params": Object {
          "BaseColor": Object {
            "value": Object {
              "a": 1,
              "b": 0.3607843220233917,
              "g": 0.7137255072593689,
              "r": 0.3490196168422699,
            },
          },
          "EmissiveStrength": Object {
            "range": Array [
              0,
              1,
            ],
            "value": 0,
          },
          "Opacity": Object {
            "range": Array [
              0,
              1,
            ],
            "value": 1,
          },
        },
        "shader": "SimpleSurfaceShader",
        "type": "Material",
      },
    },
    "Opacity": Object {
      "range": Array [
        0,
        1,
      ],
      "value": 1,
    },
    "Visible": Object {
      "value": true,
    },
  },
  "type": "GeomItem",
}
`;
